### DISCUSSION TITLE
"fetch failed" if called inside getServerSideProps with Next 13 + Node 18 + Pages directory

### DISCUSSION THREAD

**Discussion by andre-mr** [vercel/next.js#46240] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Home Single Language
    Binaries:
      Node: 18.14.1
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.1.6
      eslint-config-next: 13.0.0
      react: 18.2.0
      react-dom: 18.2.0

### Which area(s) of Next.js are affected? (leave empty if unsure)

Data fetching (gS(S)P, getInitialProps)

### Link to the code that reproduces this issue

https://github.com/andre-mr/fetch-test

### To Reproduce

- use Node 18
- create a basic next 13 app
- setup getServerSideProps with some fetch inside

### Describe the Bug

I came across the following problem: in a Next 13 project, using Node 18, with the Page directory pattern...

When I try to use a fetch inside the getServerSideProps function, it fails.
I'm running local "next dev" and requesting to a localhost api.

All works well when I use Node 16, but I have a monorepo and running 2 different node versions because of this doesn't seem to be a beautiful workaround...

I didn't find a solution here in the issues or on stackoverflow other than using node 16. it's a bug, right? 😩

### Expected Behavior

Node 18's Fetch API should work like Node 16's with Next 13, when using Pages directory and used inside getServerSideProps.

### How are you deploying your application? (if relevant)

next dev

**Comment by icyJoseph** [comment]
A couple of things...

You are trying to call:

```js
 const result = await fetch('http://localhost:3000/hello?num=10')
```

But that route, `/hello`, does not exist, you want to call `/api/hello`, I pressume.

```js
 const result = await fetch('http://localhost:3000/api/hello?num=10')
```

With that out of the way, I can already tell you that using GSSP to reach for API routes is a bit of a bad practice. Within GSSP, you are on the server, so you should just get the data you need right there.

Read more here: https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes

You'd be basically forcing your server to reach out to "the internet" back to itself... exaggerating a bit but it is an unnecessary round-trip you can perfectly avoid.

**Comment by andre-mr** [comment]
Ok, let me be clarify better then...

- That route exists. It's not an internal Next API (not "pages/api/hello"), but an external Lambda from Docker image i'm running together with Next app. In my example, port 3000 is used by that external API, while Next app mentioned is using port 3010.
- As I mentioned, when I just switch to Node 16 everything works as expected.

But I found out what is going on. It's Node >17 related, they changed default IP resolving. For most cases, "localhost" doesn't work anymore. So, for such local fetch like that, we just need to use 127.0.0.1 or other workaround to deal with new default ipv6 resolving.

https://github.com/node-fetch/node-fetch/issues/1624

**Reply by icyJoseph** [reply]
Aha, that's interesting insight, I saw someone having issues with that too, theirs was resolving to ::1:3000 - nice.

**Reply by zatarain** [reply]
Hi everyone, good evening.

I had the same problem and for my understanding I have same (or at least quite similar) setup as @andre-mr, which is following:
  * I have a docker compose swarm with:
    - Back-end container, which is my API `service` in Ruby on Rails [Ruby 3+, Rails 7+]
    - Front-end container, which is a NextJS `website` with TypeScript [NodeJS 19+, NPM 9+]
  * The `website` needs to fetch data from the `service` per request
  * So, I use `getServerSideProps` to fetch the data and populate the page props

However, I downgraded to NodeJS 16+ and it seems to be working and then moved to NodeJS 19+ and worked again. I didn't know what I did different after few hours I remembered that to move between versions I ran the `website` outside of the swarm, so in my host machine and the actual difference was that within the swarm the host name of both the `service` and the `website` were different and it uses HTTP rather than HTTPS (locally I have been using `https://api.mydomain.com.dev` and `https://mydomain.com.dev` respectively).

So, I solved the problem using an environment variable `API_URL=http://service:3000` within the docker swarm and that variable would be different on my production an staging environments like `API_URL=https://api.test.mydomain.com` and `API_URL=https://api.mydomain.com` respectively.

I hope my experience be useful.

Cheers
Uli 😊
